### PR TITLE
get_map_anns spw

### PR DESCRIPTION
**Pull request by will-moore**
This adds SPW support to the ```idr_get_map_annotations.py```, to allow transfer of map annotations from e.g. IDR to corresponding Plates/Wells on another server.
This was added in prep for OMERO.parade testing. Not urgently needed elsewhere but could be useful functionality in future.

To test:
- For PDI, this works as before (local Dataset and Image names need to match those in IDR).
E.g. copy map annotations from IDR Project:51 to local Project:203
```python idr_get_map_annotations.py username password Project:51 Project:203 --server localhost```

 - To copy map annotations from Wells in Plate:422 on IDR to local Plate:2, use:
```python idr_get_map_annotations.py username password Plate:422 Plate:2 --server localhost```
You don't need any names to match. As long as the IDR Plate has Wells existing for all the rows and columns of Wells in the local plate, this will work.

 - For local Screen, you need child Plates with names that match Plate names in the IDR Screen.
```python idr_get_map_annotations.py username password Screen:102 Screen:101 --server localhost```

### CODE REVIEW

File: maintenance/scripts/idr_get_map_annotations.py [hunk 1]
```diff
@@ -31,69 +32,128 @@
...
+    url = webclient_api_url + "plates/?id=%s" % screen_id
+    plates = session.get(url).json()['plates']
+    by_name = {}
+    for p in plates:
+        by_name[p['name']] = p
+    return by_name
+
+
+def get_idr_wells_as_grid(plate_id):
+    url = base_url + "webgateway/plate/%s/0/" % plate_id
```
Comment (jburel) [inline maintenance/scripts/idr_get_map_annotations.py:72]: one to add to the ``/api``

**Comment by jburel** [comment]
Was this script run against either outreach or workshop?
cc @pwalczysko

**Comment by pwalczysko** [comment]
> Was this script run against either outreach or workshop?

Not by myself

**Comment by will-moore** [comment]
No, but it would be a good test to do e.g.:

```
python idr_get_map_annotations.py username password Plate:422 Plate:102 --server outreach.openmicroscopy.org
```
to copy map annotations from Wells in http://idr.openmicroscopy.org/webclient/?show=plate-422 to Wells in
https://outreach.openmicroscopy.org/webclient/?show=plate-102

**Comment by joshmoore** [comment]
Just as a side note: this is starting to look less like a training script and more like generally useful functionality that could be used _in a_ training script.

**Comment by will-moore** [comment]
These scripts aren't used in the actual training. They're just used to setup the server for training sessions.

**Comment by jburel** [comment]
As mentioned this morning, I can see that going into a "trainer" guide.

**Comment by joshmoore** [comment]
Not sure where we want to draw the bar, I'd suggest having "substantial code" not as a script nor as a guide but as a pip installable artifact.

**Comment by pwalczysko** [comment]
```
(env561) pwalczysko@ls31618~$ python idr_get_map_annotations.py --server outreach.openmicroscopy.org tUSER PWd 422  102
Traceback (most recent call last):
  File "idr_get_map_annotations.py", line 179, in <module>
    main(sys.argv[1:])
  File "idr_get_map_annotations.py", line 174, in main
    args.server, args.port)
  File "idr_get_map_annotations.py", line 148, in run
    idr_id = idr_obj.split(':')[1]
IndexError: list index out of range
(env561) pwalczysko@ls31618~$ 
```

**Comment by pwalczysko** [comment]
the error https://github.com/ome/training-scripts/pull/85#issuecomment-643300048 was caused by the wrong syntax of the IDs in the command. ID is not enough, the ``Plate:`` must be there too. This was caused by misreading the following output of the help

```
(env561) pwalczysko@ls31618~$ python idr_get_map_annotations.py -h
usage: idr_get_map_annotations.py [-h] [--server SERVER] [--port PORT]
                                  username password idr_obj local_obj

positional arguments:
  username
  password
  idr_obj
  local_obj
```

Maybe add an example to help, just like in the header of this PR ?

**Comment by pwalczysko** [comment]
```
(env561) pwalczysko@ls31618~$ python idr_get_map_annotations.py --server outreach.openmicroscopy.org trainer-1 testing Dataset:51 Dataset:3018
(env561) pwalczysko@ls31618~$ 
```
Dataset is not working, but there is no warning or crash, just no output and no action

**Comment by pwalczysko** [comment]
Project is working fine

**Comment by pwalczysko** [comment]
Screen is working fine

**Comment by pwalczysko** [comment]
In summary, all is working, except 2 RFEs

1. improve help https://github.com/ome/training-scripts/pull/85#issuecomment-643304886
2. support for Datasets https://github.com/ome/training-scripts/pull/85#issuecomment-643311140

**Comment by will-moore** [comment]
I've improved the help message and the handling of invalid types "E.g. Dataset:1".
However, I don't think it's worth the extra dev time and testing effort to support `Datasets` just in case it's useful in the future. These scripts are only for our own internal use. If we need to support Datasets, we can add that at the time.

File: maintenance/scripts/idr_get_map_annotations.py [hunk 2]
```diff
@@ -158,14 +158,21 @@ def run(username, password, idr_obj, local_obj, host, port):
         conn.close()
 
 
+description = """Usage: To copy map annotations from Project: 1 on IDR
```
Comment (pwalczysko) [inline maintenance/scripts/idr_get_map_annotations.py:170]:
```suggestion
description = """Usage: To copy map annotations from Project:1 on IDR
```

**Comment by pwalczysko** [comment]
One minor correction suggest https://github.com/ome/training-scripts/pull/85/commits/87c4e2f6a5c1e3222ed42272c6709fe29c47e7de#r442324642

After that, ready to merge fmpov.

**Comment by jburel** [comment]
Few of the scripts under maintenance are now used in guides e.g. omero-guide-introduction.
We will have to rethink that repo (practical scripts have been moved)

**Pull request merged**